crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2352,16 +2352,107 @@ impl<'db> TypeInferenceBuilder<'db> {
+                            }
+                        }
+                    };
+                }
+            });
+        inferred_ty.expect("A type should always be inferred on the last comparison in the chain")
+    }
+
+    /// Infers the type of a binary comparison (e.g. '<left> == <right>'). See
+    /// `infer_compare_expression` for the higher level logic dealing with multi-comparison
+    /// expressions.
+    fn infer_binary_type_comparison(
+        &mut self,
+        left: Type<'db>,
+        op: ast::CmpOp,
+        right: Type<'db>,
+    ) -> Type<'db> {
+        match (left, right) {
+            (Type::IntLiteral(n), Type::IntLiteral(m)) => Type::BooleanLiteral(match op {
+                ast::CmpOp::Eq | ast::CmpOp::Is => n == m,
+                ast::CmpOp::NotEq | ast::CmpOp::IsNot => n != m,
+                ast::CmpOp::Lt => n < m,
+                ast::CmpOp::LtE => n <= m,
+                ast::CmpOp::Gt => n > m,
+                ast::CmpOp::GtE => n >= m,
+                // Undefined for (int, int)
+                ast::CmpOp::In | ast::CmpOp::NotIn => false,
+            }),
+            // Booleans are coded as integers (False = 0, True = 1)
+            (Type::IntLiteral(n), Type::BooleanLiteral(b)) => self.infer_binary_type_comparison(
+                Type::IntLiteral(n),
+                op,
+                Type::IntLiteral(i64::from(b)),
+            ),
+            (Type::BooleanLiteral(b), Type::IntLiteral(m)) => self.infer_binary_type_comparison(
+                Type::IntLiteral(i64::from(b)),
+                op,
+                Type::IntLiteral(m),
+            ),
+            // TODO: move this logic wherever we handle class instances
+            (Type::IntLiteral(_), Type::Instance(class_type))
+            | (Type::Instance(class_type), Type::IntLiteral(_)) => class_type
+                .is_stdlib_symbol(self.db, "builtins", "int")
+                .then(|| builtins_symbol_ty(self.db, "bool").to_instance(self.db))
+                .unwrap_or(Type::Unknown),
+            // TODO: type-generic cases (using _) should be moved after the code handling
+            // classes/instances as they can override the behaviour with literal types by
+            // implementing `__eq__`, ...
+            (Type::IntLiteral(_), _) | (_, Type::IntLiteral(_)) => match op {
+                ast::CmpOp::Eq | ast::CmpOp::Is => Type::BooleanLiteral(false),
+                ast::CmpOp::NotEq | ast::CmpOp::IsNot => Type::BooleanLiteral(true),
+                ast::CmpOp::Lt | ast::CmpOp::LtE | ast::CmpOp::Gt | ast::CmpOp::GtE => {
+                    // TODO: this fails at runtime, we might need a diagnostic
+                    Type::Unknown
+                }
+                // TODO: this is valid for some types (tuples, list, dict, set, ...)
+                ast::CmpOp::In | ast::CmpOp::NotIn => Type::Unknown,
+            },
```

Comment:
I think if we aren't going to do the full `Type::Instance` handling in this PR, we should just leave this case out entirely, as it will be entirely replaced by generic `Type::Instance` handling.

Reply:
I guess I was seeing the special case of (int, IntLiteral) as a part of IntLiteral's logic, but it does indeed fit better as a generic part of instances